app/src/main/kotlin/com/flxrs/dankchat/DankChatViewModel.kt

```diff
@@ -1,31 +1,67 @@
 package com.flxrs.dankchat
 
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
 import com.flxrs.dankchat.data.ChatRepository
 import com.flxrs.dankchat.data.DataRepository
+import com.flxrs.dankchat.data.api.ApiManager
+import com.flxrs.dankchat.preferences.DankChatPreferenceStore
+import com.flxrs.dankchat.utils.extensions.withoutOAuthSuffix
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class DankChatViewModel @Inject constructor(
     private val chatRepository: ChatRepository,
+    private val dankChatPreferenceStore: DankChatPreferenceStore,
+    private val apiManager: ApiManager,
     dataRepository: DataRepository,
 ) : ViewModel() {
 
     val commands = dataRepository.commands
     var started = false
         private set
 
+    private val _currentUserName = MutableStateFlow<String?>(null)
+    val currentUserName: StateFlow<String?> get() = _currentUserName.asStateFlow()
+
+    private val _validationError = MutableSharedFlow<OAuthValidationError>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
```

Comment:
I"ve so far always used a Channel with a unlimited buffer for these types of one-shot events, but have also heard that SharedFlows are also a valid option. Any reason why you chose them here?

Reply:
I think that the error will likely result in UI changes (like toast, or dialog) if there is too many errors (for whatever reason) there might be a lot of message showing up, so I limit it to be "latest" error only.

Reply:
That does make sense, there are conflated channels for this use case as well. Also, I think we can actually merge the username Flow and the validation error one, by having a some kind of `ValidationResult` sealed class that contains a `data class ValidUser(val userName: String)`  and the existing errors. Then we would only have a single collection for all events.